products/libllbuild/BuildDB-C-API.cpp

```diff
@@ -0,0 +1,168 @@
+
+#include <llbuild/llbuild.h>
+
+#include "llbuild/Core/BuildDB.h"
+
+#include <mutex>
+
+using namespace llbuild;
+using namespace llbuild::core;
+
+namespace {
+  
+  class CAPIBuildDBKeysResult {
+  private:
+    std::vector<KeyType> keys;
+    
+  public:
+    CAPIBuildDBKeysResult(std::vector<KeyType> keys): keys(keys) { }
+    
+    CAPIBuildDBKeysResult(const CAPIBuildDBKeysResult&) LLBUILD_DELETED_FUNCTION;
+    CAPIBuildDBKeysResult& operator=(const CAPIBuildDBKeysResult&) LLBUILD_DELETED_FUNCTION;
+    
+    const uint32_t size() {
+      return keys.size();
+    }
+    
+    const KeyType keyAtIndex(KeyID index) {
+      return keys[index];
+    }
+  };
+  
+  class CAPIBuildDB: public BuildDBDelegate {
+    /// The key table, for memory caching of key to key id mapping.
+    llvm::StringMap<KeyID> keyTable;
+    
+    /// The mutex that protects the key table.
+    std::mutex keyTableMutex;
+    
+    std::unique_ptr<BuildDB> _db;
+    
+  public:
+    CAPIBuildDB(StringRef path, uint32_t clientSchemaVersion, std::string *error_out) {
+      _db = createNondestructiveSQLiteBuildDB(path, clientSchemaVersion, error_out);
+      if (error_out != NULL) {
+        return;
+      }
+      // The delegate  caches keys for key ids
+      _db.get()->attachDelegate(this);
+      
+      if (!buildStarted(error_out)) {
+        buildComplete();
+        return;
+      }
+    }
+    
+    virtual const KeyID getKeyID(const KeyType& key) override {
+      std::lock_guard<std::mutex> guard(keyTableMutex);
+      
+      // The RHS of the mapping is actually ignored, we use the StringMap's ptr
+      // identity because it allows us to efficiently map back to the key string
+      // in `getRuleInfoForKey`.
+      auto it = keyTable.insert(std::make_pair(key, 0)).first;
+      return (KeyID)(uintptr_t)it->getKey().data();
+    }
+    
+    virtual KeyType getKeyForID(const KeyID key) override {
+      // Note that we don't need to lock `keyTable` here because the key entries
+      // themselves don't change once created.
+      return llvm::StringMapEntry<KeyID>::GetStringMapEntryFromKeyData(
+                                                                       (const char*)(uintptr_t)key).getKey();
+    }
+    
+    const bool buildStarted(std::string *error_out) {
+      return _db.get()->buildStarted(error_out);
+    }
+    
+    void buildComplete() {
+      _db.get()->buildComplete();
+    }
+    
+    const bool getKeys(std::vector<KeyType>& keys_out, std::string *error_out) {
+      return _db.get()->getKeys(keys_out, error_out);
+    }
+  };
+  
+}
+
+const llb_data_t mapData(std::vector<uint8_t> input) {
+  const auto data = input.data();
+  const auto size = sizeof(data) * input.size();
+  const auto newData = (uint8_t *)malloc(size);
+  memcpy(newData, data, size);
+  
+  return llb_data_t {
+    input.size(),
+    newData
+  };
+}
+
+const llb_database_t* llb_database_create(
```

Comment:
Should this be `create` or `open`? I think `open` makes more sense, maybe?

Reply:
It creates an instance of the database type, but opens a channel to an existing database. I think depending on the interpretation both could be right. I'm fine with changing it to open though.